### ISSUE TITLE
Faster EphemeralStream weakMemo with OCC

### ISSUE THREAD

**Issue by dansanduleac** [scalaz/scalaz#1103]
I've been running some benchmarks on `scala.Stream` vs `scalaz.EphemeralStream`.

Let's define a timing function:

``` scala
def timed(stuff: => Any, iterations: Int = 100): Unit = {
  val vs = for (_ <- 1 to iterations) yield { 
    System.gc // try to get rid of potential GC pauses
    var t = System.currentTimeMillis
    stuff
    System.currentTimeMillis - t
  }
  println(s"Average time after $iterations iterations: ${vs.sum / vs.length}ms")
}
```

Now, 

```
scala> timed(scalaz.EphemeralStream.range(0, 10000000).length)
Average time after 100 iterations: 731ms
scala> timed(Stream.range(0, 10000000).length)
Average time after 100 iterations: 418ms
```

The `EphemeralStream` traversal takes `75%` longer than the Stream traversal.

But if we were to change `EphemeralStream.weakMemo` to use an `AtomicReference` instead of a synchronized block, allowing for the head/tail to possibly be evaluated multiple times in parallel:

``` scala
def weakMemo[V](f: => V): () => V = {
    val ref: AtomicReference[WeakReference[V]] = new AtomicReference()
    () => {
      def genNew(old: WeakReference[V]) = {
        val x = f
        ref.compareAndSet(old, new WeakReference(x))
        x
      }
      val v = ref.get()
      if (v != null) {
        val crt = v.get()
        if (crt == null) {
          genNew(v)
        } else crt
      } else genNew(v)
    }
  }
```

```
scala> timed(scalaz.EphemeralStream.range(0, 10000000).length)
Average time after 100 iterations: 489ms
```

This is now only `17%` slower, yet still maintaining the benefit of `head` and `tail` being individually lazy (as opposed to, say, saving the `(head, tail)` tuple inside a `WeakReference` instead and continuing to use synchronization but for the entire tuple).

Would you say this is a worthwhile improvement?

**Comment by runarorama** [comment]
:+1:

**Comment by S11001001** [comment]
:+1:

**Comment by S11001001** [comment]
Related: #1259

**Comment by S11001001** [comment]
Fixed by #1348.
